### ISSUE TITLE
VPA : Resize each pod in a deployment

### ISSUE THREAD

**Issue by shubhambadkur** [kubernetes/autoscaler#3015]
I am trying to deploy vpa to act on each replica of the deployment not scale up and down the complete deployment. 

`apiVersion: autoscaling.k8s.io/v1beta2
kind: VerticalPodAutoscaler
metadata:
  name: {{ template "App.fullname" . }}-vpa
  namespace: {{ .Release.Namespace }}
  labels:
    {{ template "common.labels" . }}
  annotations:
    {{ template "common.annotations" . }}
spec:
  targetRef:
    apiVersion: apps/v1
    kind:       Deployment
    name:       {{ template "App.fullname" . }}
  updatePolicy:
    updateMode: Auto`
As of now VPA recommends and updates a common resource for complete deployment

**Comment by bskiba** [comment]
I'm not sure I understand the issue.

VPA recommends one value for the whole deployment, but does not modify the deployment itself, it modifies actual request of replicas (pods).

Can you tell us more on what you observed vs what you expected to happen?
Do you want each replica to have a separate recommendation?

**Comment by shubhambadkur** [comment]
@bskiba yes we want to have each replica to have separate recommendation as per our usecase.

**Comment by bskiba** [comment]
This is not a use case VPA currently supports. It is configured on a controller (e.g deployment) and recommends a common recommendation for the whole controller on the assumption that pods under one controller are "same" or at least very similar. This is in line with the fact that the deployment has a common template for pods it controls. If you specify resource requests there, they are applied to every pod. 

Can you tell me more about your use case? Why is a common recommendation not working for your deployment?

**Comment by shubhambadkur** [comment]
@bskiba our usecase is that we have essentially kafka consumers running in pods. One deployment is consuming from one topic . We try to keep the throughput across partition same but we want to scale the pod in sync with the load of the partition it is attached to.

**Comment by bskiba** [comment]
I see. Do I understand correctly that there can be only one consumer per partition?

Unfortunately the only solution that I can see now is to use a separate deployment for each partition and modify the min-replicas flag of updater to 0 so that deployments with 1 replica can be updated.
https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/main.go#L44

**Comment by bskiba** [comment]
I believe this request goes agains the philosophy behind deployments. A Deployment as a k8s concept is "cattle" - all pods should be roughly similar. You would not be able to have different requests for pods if you just used the deployment - the pod template is common. That's why I do not see implementing this feature in the near future as feasible.
/close
